test/controllers/APIControllerSpec.scala

```diff
@@ -0,0 +1,459 @@
+    }
+
+    "return product created if body follows correct structure" in {
+      val goodJson: JsValue = JsObject(
+        Seq(
+          "name" -> JsString("test"),
+          "quantity" -> JsNumber(5),
+          "price" -> JsNumber(5.00)
+        )
+      )
+      val controller = inject[APIController]
+      val correctBody = controller
+        .createProduct()
+        .apply(FakeRequest(POST, "/rest/bakery/product").withJsonBody(goodJson))
+      status(correctBody) mustBe OK
+      contentAsString(correctBody) must include("New record added")
+    }
+
+    "return product created if body is correctly structured but is oddly ordered" in {
+      val goodJson: JsValue = JsObject(
+        Seq(
+          "price" -> JsNumber(5.00),
+          "name" -> JsString("test"),
+          "quantity" -> JsNumber(5)
+        )
+      )
+      val controller = inject[APIController]
+      val correctBody = controller
+        .createProduct()
+        .apply(FakeRequest(POST, "/rest/bakery/product").withJsonBody(goodJson))
+      status(correctBody) mustBe OK
+      contentAsString(correctBody) must include("New record added")
+    }
+
+    "return product created if body contains the fields desired but also contains other data" in {
+      val extraJson: JsValue = JsObject(
+        Seq(
+          "price" -> JsNumber(5.00),
+          "name" -> JsString("test"),
+          "quantity" -> JsNumber(5),
+          "location" -> JsString("1234 W Street")
+        )
+      )
+      val controller = inject[APIController]
+      val correctBody = controller
+        .createProduct()
+        .apply(
+          FakeRequest(POST, "/rest/bakery/product").withJsonBody(extraJson)
+        )
+      status(correctBody) mustBe OK
+      contentAsString(correctBody) must include("New record added")
+    }
+  }
+
+  "APIController PUT" should {
+    "return not found if the id is not present in the db" in {
+      val okJson: JsValue = JsObject(
+        Seq(
+          "name" -> JsString("test"),
+          "price" -> JsNumber(5.00),
+          "quantity" -> JsNumber(5)
+        )
+      )
+      val controller = inject[APIController]
+      val noIdFound = controller
+        .updateProduct(UUID.fromString("00000000-0000-0000-0000-000000000000"))
+        .apply(
+          FakeRequest(PUT, "/rest/bakery/product/:id").withJsonBody(okJson)
+        )
+      status(noIdFound) mustBe NOT_FOUND
+      contentAsString(noIdFound) must include("No Product Found")
+    }
+
+    "return bad request if no body is present in the request" in {
+      val controller = inject[APIController]
+      val noBody = controller
+        .updateProduct(UUID.fromString("0a776c48-1ef0-4d2e-8fab-81d074b0c8d4"))
+        .apply(FakeRequest(PUT, "/rest/bakery/product/:id"))
+      status(noBody) mustBe BAD_REQUEST
+      contentAsString(noBody) must include("No request body found")
+    }
+
+    "return ok if an existing product was successfully updated with all 3 parameters" in {
```

Comment:
Notice how all of these tests have a pattern to them:
Whether it is bad request or ok, and the initial input is the only thing that really changes. It should be possible to do something along the lines of:

```scala
// ok request scenario
Seq(
JsObject(
        Seq(
          "name" -> JsString("test"),
          "price" -> JsNumber(5.00)
        )
      ),
JsObject(
        Seq(
          "name" -> JsString("test")
        )
      )
).foreach { testCase =>
    "return ok if an existing product was successfully updated with ${testCase.length} parameters: ${testCase.values}" in {
      val okJson: JsValue = testCase
      val controller = inject[APIController]
      val successfulUpdate = controller
        .updateProduct(UUID.fromString("0a776c48-1ef0-4d2e-8fab-81d074b0c8d4"))
        .apply(
          FakeRequest(PUT, "/rest/bakery/product/:id").withJsonBody(okJson)
        )
      status(successfulUpdate) mustBe OK
      contentAsString(successfulUpdate) must include("Record updated")
    }
}




```

Reply:
This is a stretch goal, and not required